build faild in version v20.5.2.7-stable:ISO C++17 does not allow ‘register’ storage class specifier

Make sure that `git diff` result is empty and you've just pulled fresh master. Try cleaning up cmake cache. Just in case, official build instructions are published here: https://clickhouse.yandex/docs/en/development/build/

**Operating system**
CentOS7.6

**Cmake version**
cmake 3.16.0

**Ninja version**

**Compiler name and version**
gcc-9.3.0

> Scanning dependencies of target glibc-compatibility
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/glibc-compatibility.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_divzero.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_divzerof.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_invalid.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_invalidf.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_oflow.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_oflowf.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_uflow.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_uflowf.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_xflow.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/__math_xflowf.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/clock_getres.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/clock_gettime.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/clock_nanosleep.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/exp.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/exp2.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/exp2f.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/exp2f_data.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/exp_data.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/fallocate.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/fcntl.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/futimens.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/getrandom.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/glob.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/lgamma.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/log.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/log2.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/log2_data.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/log2f.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/log2f_data.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/log_data.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/logf.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/logf_data.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/pipe2.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/posix_spawn.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/pow.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/pow_data.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/pwritev.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/sched_cpucount.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/sched_getcpu.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/syscall_ret.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/timespec_get.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/utimensat.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/vasprintf.c.o
> [  0%] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/vdso.c.o
> [  0%] Building ASM object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/aarch64/syscall.s.o
> [  0%] Building ASM object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/aarch64/longjmp.s.o
> [  0%] Linking C static library libglibc-compatibility.a
> [  0%] Built target glibc-compatibility
> Scanning dependencies of target unwind
> [  0%] Building CXX object contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/libunwind.cpp.o
> In file included from /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/UnwindCursor.hpp:69,
>                  from /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/libunwind.cpp:22:
> /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/DwarfInstructions.hpp: In instantiation of ‘static int libunwind::DwarfInstructions<A, R>::stepWithDwarf(A&, libunwind::DwarfInstructions<A, R>::pint_t, libunwind::DwarfInstructions<A, R>::pint_t, R&) [with A = libunwind::LocalAddressSpace; R = libunwind::Registers_arm64; libunwind::DwarfInstructions<A, R>::pint_t = long unsigned int]’:
> /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/UnwindCursor.hpp:933:50:   required from ‘int libunwind::UnwindCursor<A, R>::stepWithDwarfFDE() [with A = libunwind::LocalAddressSpace; R = libunwind::Registers_arm64]’
> /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/UnwindCursor.hpp:1987:18:   required from ‘int libunwind::UnwindCursor<A, R>::step() [with A = libunwind::LocalAddressSpace; R = libunwind::Registers_arm64]’
> /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/UnwindCursor.hpp:1975:5:   required from here
> /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/DwarfInstructions.hpp:223:37: error: ISO C++17 does not allow ‘register’ storage class specifier [-Werror=register]
>   223 |         register unsigned long long x17 __asm("x17") = returnAddress;
>       |                                     ^~~
> /usr/local/clickhouse/ClickHouse-20.5.2.7/contrib/libunwind/src/DwarfInstructions.hpp:224:37: error: ISO C++17 does not allow ‘register’ storage class specifier [-Werror=register]
>   224 |         register unsigned long long x16 __asm("x16") = cfa;
>       |                                     ^~~
> cc1plus: all warnings being treated as errors
> make[2]: *** [contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/libunwind.cpp.o] Error 1
> make[1]: *** [contrib/libunwind-cmake/CMakeFiles/unwind.dir/all] Error 2
> make: *** [all] Error 2
>

--- a/contrib/libunwind-cmake/CMakeLists.txt
+++ b/contrib/libunwind-cmake/CMakeLists.txt
@@ -53,7 +53,7 @@ endif ()
 # The library is using register variables that are bound to specific registers
 # Example: DwarfInstructions.hpp: register unsigned long long x16 __asm("x16") = cfa;
 check_cxx_compiler_flag(-Wregister HAVE_WARNING_REGISTER)
-if (HAVE_WARNING_MAYBE_UNINITIALIZED)
+if (HAVE_WARNING_REGISTER)
     target_compile_options(unwind PRIVATE -Wno-register)
 endif ()
 

--- a/contrib/libunwind-cmake/CMakeLists.txt
+++ b/contrib/libunwind-cmake/CMakeLists.txt
@@ -50,6 +50,13 @@ if (HAVE_WARNING_MAYBE_UNINITIALIZED)
     target_compile_options(unwind PRIVATE -Wno-maybe-uninitialized)
 endif ()
 
+# The library is using register variables that are bound to specific registers
+# Example: DwarfInstructions.hpp: register unsigned long long x16 __asm("x16") = cfa;
+check_cxx_compiler_flag(-Wregister HAVE_WARNING_REGISTER)
+if (HAVE_WARNING_MAYBE_UNINITIALIZED)
+    target_compile_options(unwind PRIVATE -Wno-register)
+endif ()
+
 install(
     TARGETS unwind
     EXPORT global

